### ISSUE TITLE
flake: h2_http_proxy_nosec_test.exe shutdown_finishes_calls  (also with h2_http_proxy_test.exe)

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#14471]
Currently our top windows flake.

Recent failures (both opt and dbg)
https://sponge.corp.google.com/invocation?id=b25f2870-793b-4c35-8efc-55e7ccfc3cd9
https://sponge.corp.google.com/invocation?id=700f9885-6094-49eb-a653-a8ab6dc9ce6e
https://sponge.corp.google.com/invocation?id=6482be2b-8ac5-4df2-9d99-4337438c9dec
https://sponge.corp.google.com/invocation?id=0c9cec20-0647-49ed-aea0-622241d1c118
https://sponge.corp.google.com/invocation?id=7c2df0f2-5aa7-471b-b703-b29a5b887a08

**Comment by jtattermusch** [comment]
```
I0220 19:50:12.860000000  3620 http_connect_handshaker.cc:298] Connecting to server localhost:10162 via HTTP proxy ipv6:[::1]:2029
D0220 19:50:12.860000000  5712 http_proxy_fixture.cc:374] on_read_request_done: 00462790 "No Error"
I0220 19:50:12.860000000  5712 subchannel.cc:605] New connected subchannel at 00436A60 for subchannel 00450B10
I0220 19:50:12.860000000  5712 http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1519185012.860000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}
D0220 19:50:12.860000000  5712 dns_resolver.cc:247] In cooldown from last resolution (from 5 ms ago). Will resolve again in 995 ms
E0220 19:50:12.860000000  5712 shutdown_finishes_calls.cc:166] assertion failed: status == GRPC_STATUS_INTERNAL || status == GRPC_STATUS_UNAVAILABLE
Abort handler called.
I0220 19:50:12.860000000  5788 http_proxy_fixture.cc:152] HTTP proxy client read: {"created":"@1519185012.860000000","description":"TCP stream shutting down","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":183,"referenced_errors":[{"created":"@1519185012.860000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}]}
D0220 19:50:12.860000000  5788 http_proxy_fixture.cc:120] endpoints: 00460230 004603E0
```

```
I0219 23:48:40.852000000  6912 subchannel.cc:605] New connected subchannel at 003975B0 for subchannel 003226D8
I0219 23:48:40.852000000  6424 http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1519112920.852000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}
D0219 23:48:40.852000000  6424 dns_resolver.cc:247] In cooldown from last resolution (from 4 ms ago). Will resolve again in 996 ms
E0219 23:48:40.852000000  6424 shutdown_finishes_calls.cc:166] assertion failed: status == GRPC_STATUS_INTERNAL || status == GRPC_STATUS_UNAVAILABLE
Abort handler called.
*** 0: 0000000000E4358F print_current_stack - 0000000000E434F0
*** 1: 0000000000E43AFB abort_handler - 0000000000E43AD0
*** 2: 0000000070A6F0CB raise - 0000000070A6ED60
*** 3: 0000000070A70622 abort - 0000000070A705F0
*** 4: 0000000000D7103A test_early_server_shutdown_finishes_inflight_calls - 0000000000D70990
```

```
I0218 15:47:36.076000000  4820 http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1518997656.076000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}
D0218 15:47:36.076000000  4820 dns_resolver.cc:247] In cooldown from last resolution (from 4 ms ago). Will resolve again in 996 ms
E0218 15:47:36.076000000  4820 shutdown_finishes_calls.cc:166] assertion failed: status == GRPC_STATUS_INTERNAL || status == GRPC_STATUS_UNAVAILABLE
Abort handler called.
```

```
I0215 12:16:47.911000000  6752 shutdown_finishes_calls.cc:37] Running test: test_early_server_shutdown_finishes_inflight_calls/chttp2/fullstack
I0215 12:16:49.942000000  6752 http_proxy_fixture.cc:530] Proxy address: localhost:18523
I0215 12:16:49.942000000  6568 http_connect_handshaker.cc:298] Connecting to server localhost:32006 via HTTP proxy ipv6:[::1]:18523
D0215 12:16:49.942000000  6752 http_proxy_fixture.cc:374] on_read_request_done: 007DFFB0 "No Error"
I0215 12:16:49.942000000  6752 subchannel.cc:605] New connected subchannel at 007DD908 for subchannel 00772F38
I0215 12:16:49.942000000  6752 http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1518725809.942000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_opt_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}
E0215 12:16:49.942000000  6752 shutdown_finishes_calls.cc:166] assertion failed: status == GRPC_STATUS_INTERNAL || status == GRPC_STATUS_UNAVAILABLE
```

```
I0212 13:49:32.012000000  5640 shutdown_finishes_calls.cc:37] Running test: test_early_server_shutdown_finishes_inflight_calls/chttp2/fullstack
I0212 13:49:34.059000000  5640 http_proxy_fixture.cc:529] Proxy address: localhost:4621
I0212 13:49:34.059000000  3468 http_connect_handshaker.cc:299] Connecting to server localhost:4956 via HTTP proxy ipv6:[::1]:4621
D0212 13:49:34.059000000  5640 http_proxy_fixture.cc:374] on_read_request_done: 0038F738 "No Error"
I0212 13:49:34.059000000  5640 subchannel.cc:605] New connected subchannel at 00381B48 for subchannel 003117C0
I0212 13:49:34.059000000  5640 http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1518472174.059000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"T:\src\github\grpc\workspace_c_windows_dbg_native\src\core\lib\iomgr\tcp_windows.cc","file_line":172}
D0212 13:49:34.059000000  5640 dns_resolver.cc:247] In cooldown from last resolution (from 4 ms ago). Will resolve again in 996 ms
E0212 13:49:34.059000000  5640 shutdown_finishes_calls.cc:166] assertion failed: status == GRPC_STATUS_INTERNAL || status == GRPC_STATUS_UNAVAILABLE
Abort handler called.
```

**Comment by jtattermusch** [comment]
I am able to reproduce locally with 
` tools/run_tests/run_tests.py -l c -c dbg -r 'h2_http_proxy_nosec_test.*shutdown_finishes_calls' -n 100`

Assert at shutdown_finishes_calls.cc:166 is failing with GRPC_STATUS_UNKNOWN (status == 2) and details "Stream removed".

**Comment by jtattermusch** [comment]
Some insight into what's happening:

- everything's going fine and server shutdown is requested and cancellation of all calls is requested https://github.com/grpc/grpc/blob/cab345efe666f03739e2ae7ef11696c486d47716/test/core/end2end/tests/shutdown_finishes_calls.cc#L154

- the backend server shuts down

- in a case where the test passes, the proxy sees an "server read" error (because the backend has shutdown)  of  `I0228 15:51:00.457000000  8320 http_proxy_fixture.cc:152] HTTP proxy server read:
 {"created":"@1519833060.457000000","description":"End of TCP stream","file":"C:\cygwin64\home\builder\grpc\src\core\lib\iomgr\tcp_windows.cc","file_line":186}` and shuts down the "client_endpoint" , which results in the actual client call to fail with expected status (INTERNAL or UNAVAILABLE).

- in a case where the test fails, the server read on the proxy fails with ` http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1519833002.546000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"C:\cygwin64\home\builder\grpc\src\core\lib\iomgr\tcp_windows.cc","file_line":174}`,
the client_endpoint of the proxy is also closed, but this time that results in client call finishing with the UNKNOWN status, which fails the test.

**Comment by jtattermusch** [comment]
@yashykt  do you have any insight into what could be happening? This flake has been plaguing us for a long time, so I'd like to finally resolve it.

**Comment by jtattermusch** [comment]
After lots of debugging, I think I have the explanation for what's happening:

1. Under normal circumstances, the server gets shutdown and the pending call is cancelled. GOAWAY and RST_STREAM is sent on the wire, it gets forwarded by the http proxy to the grpc client and the test finishes with success.

2. In the failure case, the server get shutdown correctly and the pending call is cancelled. GOAWAY and RST_STREAM is sent back to the proxy (I've verified that while debugging) and the server's tcp endpoint (winsocket) goes for shutdown.  In most cases, the GOAWAY and RST_STREAM are received by the proxy and forwarded to the client (which results in the test passing). 

3. But concurrently with the pending read from 2 (that would receive the GOAWAY and RST_STREAM frames) other things are happening: because we've just established a tcp connection, HTTP/2 ping frames are being exchanged between the client and the server (they are forwarded through the proxy).  What happens is that the client sends ping to the proxy and the proxy forwards the ping frame to the server. But if the timing is right, the attempt to write to the server happens *after* the server has requested shutdown of its TCP endpoint. It looks like that when we try to write to a socket after the peer has closed, this will end up in  WSAECONNABORTED (10053) error on our pending read (that would otherwise deliver GOAWAY and RST_STREAM, that have been correctly written by the server), and the proxy cannot forward these frames to the client and it just closes the connection with client. So the client call ends up with UNKNOWN status code instead, which fails the test.

4. Indeed, all the failures I've seen reported `http_proxy_fixture.cc:152] HTTP proxy server read: {"created":"@1519833002.546000000","description":"An established connection was aborted by the software in your host machine.\r\n","file":"C:\cygwin64\home\builder\grpc\src\core\lib\iomgr\tcp_windows.cc","file_line":174}`. The string "An established connection was aborted by the software in your host machine" corresponds to WSAECONNABORTED (10053).
